pkg/reconciler/workload/namespace/namespace_controller.go

```diff
@@ -18,119 +18,363 @@ package namespace
 
-type Controller struct {
-	queue workqueue.RateLimitingInterface
+// controller
+type controller struct {
+	queue        workqueue.RateLimitingInterface
+	enqueueAfter func(logicalcluster.Name, *corev1.Namespace, time.Duration)
+
+	kubeClusterClient kubernetesclient.ClusterInterface
+	kcpClusterClient  kcpclient.ClusterInterface
+
+	namespaceLister  corelisters.NamespaceLister
+	namespaceIndexer cache.Indexer
+
+	locationLister  schedulinglisters.LocationLister
+	locationIndexer cache.Indexer
 
-	namespaceLister corelisters.NamespaceLister
-	kubeClient      kubernetes.ClusterInterface
+	workloadClusterLister  workloadlisters.WorkloadClusterLister
+	workloadClusterIndexer cache.Indexer
+
+	placmentLister   schedulinglisters.PlacementLister
+	placementIndexer cache.Indexer
 }
 
-func (c *Controller) enqueueNamespace(obj interface{}) {
-	key, err := cache.MetaNamespaceKeyFunc(obj)
+func (c *controller) enqueueNamespace(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
 	if err != nil {
 		runtime.HandleError(err)
 		return
 	}
+	clusterName, name := clusters.SplitClusterAwareKey(key)
+
+	klog.V(2).Infof("Queueing Namespace %s|%s", clusterName.String(), name)
 	c.queue.Add(key)
 }
 
-func (c *Controller) Start(ctx context.Context, numThreads int) {
+// enqueueLocation finds placement ref to this location at first, and then namespaces bound to this placement.
+func (c *controller) enqueueLocation(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	clusterName, _ := clusters.SplitClusterAwareKey(key)
+
+	placements, err := c.placementIndexer.ByIndex(byLocationWorkspace, clusterName.String())
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	for _, obj := range placements {
+		c.enqueuePlacement(obj)
+	}
+}
+
+func (c *controller) enqueuePlacement(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	clusterName, name := clusters.SplitClusterAwareKey(key)
+
+	nss, err := c.namespaceIndexer.ByIndex(byWorkspace, clusterName.String())
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	for _, o := range nss {
+		ns := o.(*corev1.Namespace)
+		nskey := clusters.ToClusterAwareKey(logicalcluster.From(ns), ns.Name)
+		klog.V(2).Infof("Queueing namespace %s|%s because placement %s|%s changed", logicalcluster.From(ns), ns.Name, clusterName, name)
+		c.queue.Add(nskey)
+	}
+}
+
+func (c *controller) enqueueWorkloadCluster(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	clusterName, _ := clusters.SplitClusterAwareKey(key)
+
+	placements, err := c.placementIndexer.ByIndex(byLocationWorkspace, clusterName.String())
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	for _, obj := range placements {
+		c.enqueuePlacement(obj)
```

Comment:
Better follow this pattern: https://github.com/kcp-dev/kcp/pull/1420/files#diff-38618a66ae5ac372ed07887d291d8b15a58ed357bd1aa9be1f54659959ba3be4R238